File: src/leetCode/linkedList/Problem160.java
package leetCode.linkedList;

/**
 * Write a program to find the node at which the intersection of two singly linked lists begins.
 * For example, the following two linked lists:
 * <p>
 * A:       a1 → a2
 * ↘
 * c1 → c2 → c3
 * ↗
 * B:  b1 → b2 → b3
 * begin to intersect at node c1.
 * <p>
 * Notes:
 * If the two linked lists have no intersection at all, return null.
 * The linked lists must retain their original structure after the function returns.
 * You may assume there are no cycles anywhere in the entire linked structure.
 * Your code should preferably run in O(n) time and use only O(1) memory.
 * <p>
 * create by stephen on 2018/5/16
 */
public class Problem160 {
    private class ListNode {
        int val;
        ListNode next;

        ListNode(int x) {
            val = x;
        }
    }

    /**
     * 不需要知道每个链表的长度 只需要两个指针同时遍历
     * 如果A达到末尾 就重新跳转到B的head。两者相遇的地方就是交点
     */
    public ListNode getIntersectionNode(ListNode headA, ListNode headB) {
        if (headA == null || headB == null) return null;
        ListNode a = headA, b = headB;
        while (a != b) {
            a = a == null ? headB : a.next;
            b = b == null ? headA : b.next;
        }

        return a;
    }

    /**
     * 找到链表的长度 然后从长度一样的地方一起往下遍历
     */
    public ListNode getIntersectionNode2(ListNode headA, ListNode headB) {
        int lenA = length(headA), lenB = length(headB);
        // move headA and headB to the same start point
        while (lenA > lenB) {
            headA = headA.next;
            lenA--;
        }
        while (lenA < lenB) {
            headB = headB.next;
            lenB--;
        }
        // find the intersection until end
        while (headA != headB) {
            headA = headA.next;
            headB = headB.next;
        }
        return headA;
    }

    private int length(ListNode node) {
        int length = 0;
        while (node != null) {
            node = node.next;
            length++;
        }
        return length;
    }

}

File: leetCode/linkedList/Problem160.cpp
#include <algorithm>

#include "list.h"

/**
 * solution 1:
 * find the length of list, then start traverse from the point where
 * the length to the end is the same
 * O(n) time + O(1) space
 */
ListNode *getIntersectionNode1(ListNode *headA, ListNode *headB) {
  int lena = 0, lenb = 0;
  ListNode *pa = headA, *pb = headB;
  while (pa) {
    pa = pa->next;
    lena++;
  }
  while (pb) {
    pb = pb->next;
    lenb++;
  }

  if (lena < lenb) {
    std::swap(lena, lenb);
    std::swap(headA, headB);
  }

  pa = headA;
  pb = headB;
  for (int i = 0; i < lena - lenb; ++i) {
    pa = pa->next;
  }

  while (pa) {
    if (pa == pb) return pa;
    pa = pa->next;
    pb = pb->next;
  }
  return nullptr;
}

/**
 * two pointer, pointer 1 traverse headA then headB,
 * pointer 2 traverse headB then headA.
 * O(n) time + O(1) space
 */
ListNode *getIntersectionNode2(ListNode *headA, ListNode *headB) {
  ListNode *pa = headA, *pb = headB;
  // bool changea = false, changeb = false;
  //   while (pa && pb) {
  //     if (pa == pb) return pa;
  //     pa = pa->next;
  //     if (!pa && !changea) {
  //       pa = headB;
  //       changea = true;
  //     }

  //     pb = pb->next;
  //     if (!pb && !changeb) {
  //       pb = headA;
  //       changeb = true;
  //     }
  //   }
  while (pa != pb) {
    pa = pa->next;
    pb = pb->next;
    if (pa == pb) return pa; // note: handle nullptr case here
    if (!pa) pa = headB;
    if (!pb) pb = headA;
  }
  return pa;
}